### PR TITLE
Olympus V2 Zap In

### PR DESCRIPTION
**Pull request by Zayen-X**
-needs peer review
-TODO in CheapestBondHelper

### CODE REVIEW

File: contracts/zaps/OlympusZapV2.sol [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            return gOHMRec;
+        }
+        IStaking(staking).stake(msg.sender, amount, false, false);
+        IStaking(staking).claim(msg.sender, false);
+        return amount;
+    }
+
+    function _exitOlympus(address fromToken, uint256 amount) internal returns (uint256) {
+        if (fromToken == gOHM) {
+            uint256 sOHMRec = IStaking(staking).unwrap(address(this), amount);
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol]: You don't need to unwrap. the unstake method takes in a `rebase` param that determins if its gohm or not

https://github.com/OlympusDAO/olympus-contracts/blob/main/contracts/Staking.sol#L170-L191

File: contracts/zaps/OlympusZapV2.sol [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        sOHM = _sOHM;
+    }
+
+    /// @notice update state for gOHM
+    function update_gOHM(address _gOHM) external onlyOlympusDAO {
+        gOHM = _gOHM;
+    }
+
+    /// @notice update state for gOHM
+    function update_approvals(
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol]: What is this used for?

File: contracts/zaps/CheapestBondHelper.sol [hunk 3]
```diff
@@ -1,56 +1,71 @@
...
-    function principle() external view returns (address);
-}
+import "./interfaces/IBondDepoV2.sol";
 
 contract CheapestBondHelper {
+    ////////////////////////// STORAGE //////////////////////////
 
-    //////////////// State ////////////////
+    /// @notice used for access control
+    address public olympusDAO = msg.sender;
```
Comment (Zayen-X) [inline contracts/zaps/CheapestBondHelper.sol]: whats the diff between this and setting it in the constructor?

File: contracts/zaps/OlympusZapV2.sol [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        emit zapOut(msg.sender, toToken, tokensRec, affiliate);
+    }
+
+    ////////////////////////// INTERNAL //////////////////////////
+
+    function _enterOlympus(uint256 amount, address toToken) internal returns (uint256) {
+        if (toToken == gOHM) {
+            IStaking(staking).stake(address(this), amount, false, false);
+            IStaking(staking).claim(address(this), false);
+            uint256 gOHMRec = IStaking(staking).wrap(msg.sender, amount);
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol]: same thing here. don't need to wrap. stake takes in a `rebase` boolean:
https://github.com/OlympusDAO/olympus-contracts/blob/main/contracts/Staking.sol#L252-L264

File: contracts/zaps/OlympusZapV2.sol [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        depo = _depo;
+    }
+
+    /// @notice update state for OHM
+    function update_OHM(address _OHM) external onlyOlympusDAO {
+        OHM = _OHM;
+    }
+
+    /// @notice update state for sOHM
+    function update_sOHM(address _sOHM) external onlyOlympusDAO {
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol]: don't think we need token update methods.

Comment (FinagleLord) [reply contracts/zaps/OlympusZapV2.sol]: update token functions have been removed, tokens are now immutable

File: contracts/zaps/CheapestBondHelper.sol (hunk 3 above)
Comment (FinagleLord) [reply contracts/zaps/CheapestBondHelper.sol]: there isn't one, but for the sake of tidiness I moved it to the constructor 🙂

File: contracts/zaps/OlympusZapV2.sol (hunk 2 above)
Comment (FinagleLord) [reply contracts/zaps/OlympusZapV2.sol]: Its used for approving the depo for new principals, I also added an _approvalCheck() before any transferFrom is called from the zap to allow the zap to function in the event that an approval is too low.

NOTE: the comment description has also been fixed.

File: contracts/zaps/OlympusZapV2.sol (hunk 4 above)
Comment (FinagleLord) [reply contracts/zaps/OlympusZapV2.sol]: removed, and fixed params

File: contracts/zaps/OlympusZapV2.sol (hunk 1 above)
Comment (FinagleLord) [reply contracts/zaps/OlympusZapV2.sol]: removed, and fixed params

File: contracts/zaps/OlympusZapV2.sol [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    /// @param toToken Either sOHM or gOHM
+    /// @return OHMRec quantity of sOHM or gOHM  received (depending on toToken)
+    function _stake(uint256 amount, address toToken) internal returns (uint256) {
+        uint256 claimedTokens;
+        // approve staking for OHM if needed
+        _approveToken(OHM, staking, amount);
+
+        if (toToken == gOHM) {
+            // stake OHM -> gOHM
+            IStaking(staking).stake(address(this), amount, false, false);
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol]: The last param of this method is `_claim` which removes the need for the additional call in the next line. You can do this for the sOHM call as well.

File: contracts/zaps/OlympusZapV2.sol [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    /// @param swapData DEX swap data
+    /// @param referral The front end operator address
+    /// @return OHMRec The quantity of sOHM or gOHM received (depending on toToken)
+    function ZapStake(
+        address fromToken,
+        uint256 amountIn,
+        address toToken,
+        uint256 minToToken,
+        address swapTarget,
+        bytes calldata swapData,
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol:74]: Where are docs/codes on how we're generating this?

File: contracts/zaps/OlympusZapV2.sol [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        address referral
```
Comment (Zayen-X) [inline contracts/zaps/OlympusZapV2.sol:75]: Whats this used for?

File: test/zaps/OlympusZapV2.test.ts [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+
+import BondId from "../../libs/constants/bondId";
+
+chai.use(solidity);
+const { expect } = chai;
+
+const OlympusZapArtifact = "Olympus_V2_Zap_In";
+
+describe("OlympusDAO Zap", () => {
```
Comment (Zayen-X) [inline test/zaps/OlympusZapV2.test.ts:24]: Thank you so much for putting this together. 🙏

File: libs/exchange/exchange.helper.ts [hunk 10]
```diff
@@ -30,3 +30,28 @@ export const exchangeAndApprove = async (
 
   return await token.balanceOf(await fromAccount.getAddress());
 };
+
+export const exchange = async (
```
Comment (Zayen-X) [inline libs/exchange/exchange.helper.ts:34]: I think the file name is weird.

**Review by Zayen-X — COMMENTED** [review]
Initial pass at solidity. Haven't looked at tests.

**Pull request merged**